lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HdfsFileSystem.java

```diff
@@ -188,4 +191,46 @@ public FileIterator listFiles(Location location)
             }
         });
     }
+
+    @Override
+    public Optional<Boolean> directoryExists(Location location)
+            throws IOException
+    {
+        stats.getDirectoryExistsCalls().newCall();
+        Path directory = hadoopPath(location);
+        FileSystem fileSystem = environment.getFileSystem(context, directory);
+
+        if (!hierarchical(fileSystem, location)) {
+            return Optional.empty();
+        }
+
+        return Optional.of(environment.doAs(context.getIdentity(), () -> {
+            try (TimeStat.BlockTimer ignored = stats.getDirectoryExistsCalls().time()) {
+                FileStatus fileStatus = fileSystem.getFileStatus(directory);
+                return fileStatus.isDirectory();
+            }
+            catch (FileNotFoundException e) {
+                return false;
+            }
+            catch (IOException e) {
+                stats.getListFilesCalls().recordException(e);
+                throw e;
+            }
+        }));
+    }
+
+    private boolean hierarchical(FileSystem fileSystem, Location rootLocation)
```

Comment:
This is clever, but expensive. Should we cache this? We could use an identity map based on the file system, since they are singletons.

Reply:
Good point, that's probably worth doing.